subsys/llext/llext_mem.c

```diff
@@ -68,18 +68,32 @@ static int llext_copy_section(struct llext_loader *ldr, struct llext *ext,
 	}
 	ext->mem_size[mem_idx] = ldr->sects[mem_idx].sh_size;
 
-	if (ldr->sects[mem_idx].sh_type != SHT_NOBITS &&
-	    IS_ENABLED(CONFIG_LLEXT_STORAGE_WRITABLE)) {
-		/* Directly use data from the ELF buffer if peek() is supported */
-		ext->mem[mem_idx] = llext_peek(ldr, ldr->sects[mem_idx].sh_offset);
-		if (ext->mem[mem_idx]) {
-			llext_init_mem_part(ext, mem_idx, (uintptr_t)ext->mem[mem_idx],
-				ldr->sects[mem_idx].sh_size);
+	if (IS_ENABLED(CONFIG_LLEXT_STORAGE_WRITABLE)) {
+		if (ldr->sects[mem_idx].sh_type != SHT_NOBITS) {
+			/* Directly use data from the ELF buffer if peek() is supported */
+			ext->mem[mem_idx] = llext_peek(ldr, ldr->sects[mem_idx].sh_offset);
+			if (ext->mem[mem_idx]) {
+				llext_init_mem_part(ext, mem_idx, (uintptr_t)ext->mem[mem_idx],
+						    ldr->sects[mem_idx].sh_size);
+				ext->mem_on_heap[mem_idx] = false;
+				return 0;
+			}
+		} else if (ldr_parm->pre_located) {
+			/*
+			 * ldr_parm is NULL only for LLEXT_MEM_SHSTRTAB and LLEXT_MEM_STRTAB
+			 * and with them sht_type won't be equal to SHT_NOBITS so we enter the
+			 * above "if" branch, not here
+			 */
```

Comment:
Understood - but if you ever end up looking at that comment, it means you wasted some time debugging an issue where someone improperly called this function. The cost of making it future proof is a single assembly instruction, it looks affordable to me :slightly_smiling_face: 
```suggestion
		} else if (ldr_parm && ldr_parm->pre_located) {
```

Reply:
@pillo79 to me "impossible" code is an indication that the author didn't understand something. So even if adding that check, I'd have to add a comment like "yes, currently the check is redundant, but we put it here for the reviewer's peace of mind" ;-)